### PR TITLE
Use normal PrototypeFactory in Robolectric tests

### PR DESCRIPTION
**Pull request by phillipm**
Make sure Robolectric tests that use `CommCareTestApplication` share more prototype factory code with the live app.

Would have caught the bug fixed by https://github.com/dimagi/commcare-odk/pull/1215 

Includes commits from https://github.com/dimagi/commcare-odk/pull/1224

### CODE REVIEW

File: app/src/org/commcare/CommCareApplication.java [hunk 1]
```diff
@@ -1476,4 +1478,7 @@ public boolean checkPendingBuildRefresh() {
         return false;
     }
 
+    public PrototypeFactory getPrototypeFactory(Context c) {
```
Comment (wpride) [inline app/src/org/commcare/CommCareApplication.java]: Why is this function necessary? (instead of just calling DbUtil.getPrototypeFactory()

Comment (phillipm) [inline app/src/org/commcare/CommCareApplication.java]: To allow for overriding in the robolectric tests via `CommCareTestApplication`

Comment (wpride) [inline app/src/org/commcare/CommCareApplication.java]: Interesting - and you cant override static methods I'm guessing?

Having to get the singleton CommCareApplication to call this method that doesn't actually need the instance seems wrong to me. Am I missing something or is this just a tradeoff we have to accept to have tests?

File: app/src/org/commcare/interfaces/HttpRequestEndpoints.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Date;
+
+/**
+ * Types of http requests made by CommCare mobile to server
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public interface HttpRequestEndpoints {
```
Comment (wpride) [inline app/src/org/commcare/interfaces/HttpRequestEndpoints.java]: Very cool

File: app/src/org/commcare/models/database/DbUtil.java [hunk 3]
```diff
@@ -28,6 +28,7 @@
     public final static String orphanFileTableName = "OrphanedFiles";
 
     private static PrototypeFactory factory;
+    private static final String[] packageNames = new String[]{"org.javarosa", "org.commcare", "org.odk.collect"};
```
Comment (wpride) [inline app/src/org/commcare/models/database/DbUtil.java]: org.odk.collect - can be removed yeah?

File: app/src/org/commcare/CommCareApplication.java (hunk 1 above)
Comment (phillipm) [inline app/src/org/commcare/CommCareApplication.java]: It does feel wrong and there are various tools to get around it, like dependency injection frameworks.

I would like to revisit this at some point, but for now it is the best way I could come up with.

And yeah, overriding static methods isn't a thing in Java.

File: app/src/org/commcare/models/database/DbUtil.java (hunk 3 above)
Comment (phillipm) [inline app/src/org/commcare/models/database/DbUtil.java]: We had to add it back due to not currently having the ability to migrate the keys in a prototype factory. I have an idea of how to do that, and will eventually follow through and remove this.

**Pull request merged**